workloads/Haskell/LuParser/src/LuParser.hs

```diff
@@ -0,0 +1,298 @@
+    -}
+
+stringP :: String -> Parser ()
+{-! -}
+stringP s = () <$ wsP (P.string s)
+{-!! stringP_1 -}
+{-!
+stringP s = () <$ P.string s
+-}
+
+constP :: String -> a -> Parser a
+constP s v = v <$ stringP s
+
+parens :: Parser a -> Parser a
+parens x = P.between (stringP "(") x (stringP ")")
+
+braces :: Parser a -> Parser a
+braces x = P.between (stringP "{") x (stringP "}")
+
+brackets :: Parser a -> Parser a
+brackets x = P.between (stringP "[") x (stringP "]")
+
+valueP :: Parser Value
+valueP = intValP <|> boolValP <|> nilValP <|> stringValP
+
+intValP :: Parser Value
+intValP = IntVal <$> wsP P.int
+
+boolValP :: Parser Value
+{-! -}
+boolValP = (\s -> BoolVal True) <$> wsP (P.string "true") <|> (\s -> BoolVal False) <$> wsP (P.string "false")
+{-!! boolValP_1 -}
+{-!
+boolValP = (\s -> BoolVal True) <$> (P.string "true") <|> (\s -> BoolVal False) <$> (P.string "false")
+-}
+
+nilValP :: Parser Value
+nilValP = NilVal <$ wsP (P.string "nil")
+
+stringValP :: Parser Value
+{-! -}
+stringValP = StringVal <$> quotes (some any <|> pure [])
+{-!! stringValP_1 -}
+{-!
+stringValP = StringVal <$> quotes (some any)
+-}
+{-!! stringValP_2 -}
+{-!
+stringValP = StringVal <$> quotes (some P.alpha)
+-}
+{-!! stringValP_3 -}
+{-!
+stringValP = StringVal <$> (quotes (some any) <|> pure [])
+-}
+  where
+    quotes :: Parser a -> Parser a
+    quotes x = P.between (P.string "\"") x (stringP "\"")
+    any :: Parser Char
+    any =
+    {-! -}
+      P.satisfy $ not . eqChar '\"'
+    {-!! stringValP_4 -}
+    {-!
+      P.satisfy (const True)
+    -}
+
+expP :: Parser Expression
+expP = compP
+  where
+    compP = catP `P.chainl1` opAtLevel (level Gt)
+    catP = sumP `P.chainl1` opAtLevel (level Concat)
+    sumP = prodP `P.chainl1` opAtLevel (level Plus)
+    prodP = uopexpP `P.chainl1` opAtLevel (level Times)
+    uopexpP =
+      baseP
+        <|> Op1 <$> uopP <*> uopexpP
+    baseP =
+      tableConstP
+        <|> Var <$> varP
+        <|> parens expP
+        <|> Val <$> valueP
+
+opAtLevel :: Int -> Parser (Expression -> Expression -> Expression)
+opAtLevel l = flip Op2 <$> P.filter (\x -> level x == l) bopP
+
+varP :: Parser Var
+varP = mkVar <$> prefixP <*> some indexP <|> (Name . VarName <$> nameP)
+  where
+    mkVar :: Expression -> [Expression -> Var] -> Var
+    mkVar e l = foldr1 (\f p u -> p (Var (f u))) l e
+
+    prefixP :: Parser Expression
+    prefixP = parens expP <|> (Var . Name . VarName <$> nameP)
+
+    indexP :: Parser (Expression -> Var)
+    indexP =
+      flip Dot <$> (P.string "." *> (VarName <$> nameP))
+        <|> flip Proj <$> brackets expP
+
+
+reserved :: [String]
+{-! -}
+reserved =
+  [ "and",
+    "break",
+    "do",
+    "else",
+    "elseif",
+    "end",
+    "false",
+    "for",
+    "function",
+    "goto",
+    "if",
+    "in",
+    "local",
+    "nil",
+    "not",
+    "or",
+    "repeat",
+    "return",
+    "then",
+    "true",
+    "until",
+    "while"
+  ]
+{-!! reserved_1 -}
```

Comment:
I'm curious how you chose these lists. Is there a significant difference in inputs to failure if you just, say, don't reserve any words? Are there certain words where removing them from the reserved list is particularly hard to catch?

Reply:
I designed mutant `reserved_1` to exclude loop-related reserved words and `reserved_2` to exclude all boolean operators/ values from the list. It timed out before catching `reserved_1`, but it took 8 inputs to catch `reserved_2`, probably because true and false values are generated and are mistaken as variable names when they are not in the reserved word list. For the other reserve words, they will only be caught when random strings coincidentally match them.

Reply:
This is really nice information! Let's figure out a way to include it with the new workload?